perf: Optimize column_encryption_policy checks in recv_results_rows (1.7x speedup for Cython, 1.2x for Python)

mykaul opened the pull request:
Optimize column_encryption_policy checks in the result parsing hot path.

Previously, every decoded value checked `ce_policy and ce_policy.contains_column(coldesc)` even when no encryption policy was configured (the common case). This PR splits the parsing into two code paths — `unpack_plain_row()` and `unpack_col_encrypted_row()` — and branches once per result set instead of per value.

Changes:
- **protocol.py**: Split `recv_results_rows()` pure-Python path
- **obj_parser.pyx**: Split `TupleRowParser.unpack_row()` into `unpack_plain_row()` / `unpack_col_encrypted_row()`, with `@boundscheck(False)` / `@wraparound(False)` and try/except moved outside the per-column loop
- **ListParser / LazyParser / row_parser**: Branch once on `column_encryption_policy`
- **numpy_parser.pyx**: `NumpyParser.parse_rows()` raises `NotImplementedError` when column encryption is configured (it has no decryption support)
- **query.py**: Same split-path optimization in `BoundStatement.bind()`

## Benchmark (Cython ListParser, 10k rows, no encryption policy)

```
                      Master            PR            Speedup
2 cols (Int32+UTF8):  3.67 ms           2.08 ms       1.77x  (2.7M -> 4.8M rows/s)
10 cols (all Int32):  8.37 ms           5.18 ms       1.62x  (1.2M -> 1.9M rows/s)
```

Improvement comes from eliminating per-value overhead in the hot loop:
- `ce_policy and ce_policy.contains_column(coldesc)` truthiness check
- `coldesc = desc.coldescs[i]` lookup (not needed in plain path)
- `try/except` per value moved outside the loop
- `@cython.boundscheck(False)` / `@cython.wraparound(False)` decorators

Fixes: https://github.com/scylladb/python-driver/issues/582

## Pre-review checklist

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [ ] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

dkropachev reviewed (changes requested):
numpy_parser.pyx, parsing.pyx also needs to be fixed

dkropachev commented on cassandra/obj_parser.pyx:
```diff
@@ -84,11 +92,36 @@ cdef class TupleRowParser(RowParser):
+
+        return res
+
+    cpdef unpack_row(self, BytesIOReader reader, ParseDesc desc):
```
You have changed API the following way:
TupleRowParser.unpack_row (existing API) -> unpacking only plain row
TupleRowParser.unpack_ce_row (new API) -> unpacking only encrypted row

The problem is that there could be dependency on existing API `TupleRowParser.unpack_row` and users that uses this API for encrypted row will have regression.

You need to keep old API cover both cases of encrypted and plain rows, and introduce two more APIs for encrypted and plain cases.

dkropachev commented:
> > numpy_parser.pyx, parsing.pyx also needs to be fixed
> 
> It wasn't clear to me what needs to fixing there: numpy - it doesn't use/call column_encryption parsing.pyx - there's just 'not implemented' functions there.

Instead of changing only `TupleRowParser` class you need to do similar changes to `ColumnParser` and `RowParser` and all their implementations: `ListParser`, `LazyParser` and `NumpyParser`, you will discover that `NumpyParser` disregards `column_encryption_policy` which means that it does not work properly when when columns are encrypted, have a plug there that throwing `NotImplementedError`.
In total all Parser implementation classes should contain implementations for:
```
    cpdef parse_plain_rows(self, BytesIOReader reader, ParseDesc desc)

    cpdef parse_enc_rows(self, BytesIOReader reader, ParseDesc desc)
```

While Abstract classes:
```
cdef class ColumnParser:
    cpdef parse_rows(self, BytesIOReader reader, ParseDesc desc):
        if desc.column_encryption_policy:
            return self.unpack_enc_row(reader, desc)
        return self.unpack_plain_row(reader, desc)

    cpdef parse_plain_rows(self, BytesIOReader reader, ParseDesc desc)

    cpdef parse_enc_rows(self, BytesIOReader reader, ParseDesc desc)
```

mykaul committed:
Optimize column_encryption_policy checks in recv_results_rows

Split recv_results_rows() into two code paths: one for when
column_encryption_policy is set and one for the common case without it.
This avoids the per-value 'ce_policy and ce_policy.contains_column(...)'
check, which is redundant for the vast majority of queries.

mykaul committed:
Optimize column_encryption_policy checks in recv_results_rows

Split recv_results_rows() into two code paths: one for when
column_encryption_policy is set and one for the common case without it.
This avoids the per-value 'ce_policy and ce_policy.contains_column(...)'
check, which is redundant for the vast majority of queries.

mykaul committed:
Optimize column_encryption_policy checks in Cython parsers

Split TupleRowParser.unpack_row() into unpack_plain_row() and
unpack_col_encrypted_row(), branching once in the callers (ListParser,
LazyParser, row_parser error path) rather than per value.

- Add boundscheck/wraparound Cython optimizations on both methods
- Extract try/except outside the per-column loop
- NumpyParser.parse_rows() raises NotImplementedError when column
  encryption is configured, since it has no decryption support
- Update parsing.pxd declarations to match

mykaul committed:
Optimize column_encryption_policy checks in Cython parsers

Split TupleRowParser.unpack_row() into unpack_plain_row() and
unpack_col_encrypted_row(), branching once in the callers (ListParser,
LazyParser, row_parser error path) rather than per value.

- Add boundscheck/wraparound Cython optimizations on both methods
- Extract try/except outside the per-column loop
- NumpyParser.parse_rows() raises NotImplementedError when column
  encryption is configured, since it has no decryption support
- Update parsing.pxd declarations to match

mykaul committed:
Optimize column_encryption_policy checks in BoundStatement.bind()

Apply the same split-path optimization to the bind() method, checking
for column_encryption_policy once rather than per parameter.

mykaul committed:
Optimize column_encryption_policy checks in BoundStatement.bind()

Apply the same split-path optimization to the bind() method, checking
for column_encryption_policy once rather than per parameter.

mykaul committed:
Add unit tests for column_encryption_policy optimization

Cover both the pure-Python path (ResultMessage.recv_results_rows) and
the Cython fast path (ListParser, NumpyParser) with and without
column encryption policy.

mykaul committed:
Add unit tests for column_encryption_policy optimization

Cover both the pure-Python path (ResultMessage.recv_results_rows) and
the Cython fast path (ListParser, NumpyParser) with and without
column encryption policy.

mykaul committed:
perf: skip ColDesc namedtuple allocation when no column encryption policy

When column_encryption_policy is None (the common case), the ColDesc
namedtuple list is never accessed by unpack_plain_row(). Skip building
it entirely, passing None to ParseDesc instead.

This avoids N namedtuple allocations per query (where N = number of
columns) on the Cython fast path for non-encrypted workloads.

mykaul committed:
perf: skip ColDesc namedtuple allocation when no column encryption policy

When column_encryption_policy is None (the common case), the ColDesc
namedtuple list is never accessed by unpack_plain_row(). Skip building
it entirely, passing None to ParseDesc instead.

This avoids N namedtuple allocations per query (where N = number of
columns) on the Cython fast path for non-encrypted workloads.

mykaul commented:
## Rebased to origin/master + Pure-Python benchmark

Rebased cleanly onto current `origin/master` (no conflicts). All 16 protocol unit tests + 112 other unit tests pass.

### Pure-Python recv_results_rows decode benchmark

The original PR description covers Cython ListParser performance. Here are the pure-Python path numbers (10,000 rows, no `column_encryption_policy`, best of 5 runs, CPU pinned to single core):

```
Config                        Master (ms)    PR (ms)  Speedup   Master rows/s       PR rows/s
----------------------------------------------------------------------------------------------------
2 cols (Int32+UTF8)               15.61 ms    12.78 ms    1.22x       640,768/s       782,702/s
10 cols (Int32+UTF8)              41.32 ms    34.21 ms    1.21x       241,999/s       292,289/s
2 cols (all Int32)                13.54 ms    11.26 ms    1.20x       738,461/s       888,109/s
10 cols (all Int32)               41.01 ms    34.46 ms    1.19x       243,837/s       290,209/s
```

Consistent **1.19–1.22x speedup** on the pure-Python path by eliminating:
- `ColDesc` namedtuple allocation per column (N allocations per query)
- Per-value `column_encryption_policy and ce_policy.contains_column(coldesc)` truthiness check
- Per-value `col_desc` lookup from the `col_descs` list
- 3-way `zip(row, column_metadata, col_descs)` → 2-way `zip(row, column_metadata)`

Environment: Python 3.14.3, no Cython `.so` for protocol.py.

mykaul commented:
## Updated benchmarks — Cython and Pure-Python paths

All benchmarks: 10,000 rows, no `column_encryption_policy` (common case), best of 5 runs, CPU pinned to single core (core 15), 1000 iters (Cython) / 200 iters (pure-Python), with warm-up. Python 3.14.3, Cython 3.2.4.

### Cython ListParser (obj_parser.pyx + row_parser.pyx)

```
Config                       Master ns/col-row    PR ns/col-row  Speedup
-------------------------------------------------------------------------
2 cols (Int32+UTF8)                 91.4 ns           81.2 ns     1.13x
10 cols (Int32+UTF8)                88.0 ns           74.1 ns     1.19x
2 cols (all Int32)                  58.6 ns           45.1 ns     1.30x
10 cols (all Int32)                 38.5 ns           28.4 ns     1.36x
```

### Pure-Python (protocol.py recv_results_rows)

```
Config                       Master ns/col-row    PR ns/col-row  Speedup
-------------------------------------------------------------------------
2 cols (Int32+UTF8)                788.7 ns          649.4 ns     1.21x
10 cols (Int32+UTF8)               432.5 ns          355.1 ns     1.22x
2 cols (all Int32)                 680.8 ns          564.5 ns     1.21x
10 cols (all Int32)                449.0 ns          358.2 ns     1.25x
```

### Summary

| Path | Speedup range |
|------|--------------|
| Cython ListParser | **1.13–1.36x** |
| Pure-Python decode | **1.21–1.25x** |

Cython gains are larger for Int32-only workloads (up to 1.36x) because the per-column overhead eliminated by this PR is a bigger fraction of the total when the deserialization itself is cheap (Int32 = single `struct.unpack`). Mixed Int32+UTF8 workloads show a smaller but still significant improvement.

mykaul commented on tests/unit/test_protocol.py:
```diff
@@ -555,3 +558,250 @@ def test_frames_without_features(self):
+                        "contains_column should be called for each value when policy exists")
+
+
+class CythonParserTest(unittest.TestCase):
```
Already addressed in an earlier push: `CythonParserTest` in `tests/unit/test_protocol.py` is decorated with `@cythontest` (imported from `tests.unit.cython.utils`), so the class is skipped when the Cython extensions are not built, instead of failing with `ImportError`. Verified this is present on the current `remove_enc_cols` branch tip. Resolving.

mykaul commented on cassandra/numpy_parser.pyx:
```diff
@@ -81,6 +81,10 @@ cdef class NumpyParser(ColumnParser):
         cdef ArrDesc[::1] array_descs
         cdef ArrDesc *arrs
 
+        if desc.column_encryption_policy:
+            raise NotImplementedError(
+                "NumpyParser does not support column encryption")
```
Fixed. In `cassandra/row_parser.pyx`, the `except Exception as e:` wrapping `colparser.parse_rows(reader, desc)` now has a preceding `except NotImplementedError: raise` so the NumPy+column-encryption rejection propagates to the caller instead of being caught and silently retried with `TupleRowParser`. All other exception types still fall back to `TupleRowParser` as before.

Also rewrote `test_numpy_parser_rejects_encryption` (tests/unit/test_protocol.py) to exercise the real production path via `make_recv_results_rows(NumpyParser())` bound to a `ResultMessage` subclass, instead of calling `NumpyParser().parse_rows()` directly - this is the same wrapper `NumpyProtocolHandler` uses. Verified the rewritten test fails against the pre-fix wrapper (a `DriverException` from the fallback surfaces instead of `NotImplementedError`) and passes with the fix applied.

mykaul commented on tests/unit/test_protocol.py:
```diff
@@ -555,3 +558,250 @@ def test_frames_without_features(self):
+
+        msg.recv_results_rows(f, ProtocolVersion.V4, {}, None, mock_policy)
+
+        # With optimization: policy existence checked once, contains_column called per value
+        # = 100 rows * 2 columns = 200 calls to contains_column
+        # The key is we avoid checking 'column_encryption_policy and ...' 200 times
+        self.assertEqual(mock_policy.contains_column.call_count, 200,
+                        "contains_column should be called for each value when policy exists")
```
Fixed exactly as suggested. Added a `_BoolCountingPolicy` fake policy whose `__bool__` increments a counter, used it in `test_optimization_efficiency`, and now assert `bool_call_count == 1` (policy truthiness checked once per result message) in addition to the existing `contains_column_call_count == 200` check. Verified this actually catches the regression this test was meant to guard against: temporarily reintroducing the old per-value `column_encryption_policy and column_encryption_policy.contains_column(...)` check in `cassandra/protocol.py` makes `bool_call_count` come back as 200 instead of 1, so the rewritten test fails (the old assertion alone would still have passed, which was the bug).

mykaul commented on cassandra/obj_parser.pyx:
```diff
@@ -47,48 +51,78 @@ def parse_rows_lazy(BytesIOReader reader, ParseDesc desc):
-
-            # Deserialize bytes to python object
-            deserializer = desc.deserializers[i]
-            coldesc = desc.coldescs[i]
-            uses_ce = ce_policy and ce_policy.contains_column(coldesc)
-            try:
+        try:
+            for i in range(rowsize):
+                # Read the next few bytes
+                get_buf(reader, &buf)
+
+                # Deserialize bytes to python object
+                deserializer = desc.deserializers[i]
+                coldesc = desc.coldescs[i]
+                uses_ce = ce_policy.contains_column(coldesc)
                 if uses_ce:
                     col_type = ce_policy.column_type(coldesc)
                     decrypted_bytes = ce_policy.decrypt(coldesc, to_bytes(&buf))
                     PyBytes_AsStringAndSize(decrypted_bytes, &newbuf.ptr, &newbuf.size)
-                    deserializer = find_deserializer(ce_policy.column_type(coldesc))
+                    deserializer = find_deserializer(col_type)
                     val = from_binary(deserializer, &newbuf, desc.protocol_version)
                 else:
                     val = from_binary(deserializer, &buf, desc.protocol_version)
-            except Exception as e:
-                raise DriverException('Failed decoding result column "%s" of type %s: %s' % (desc.colnames[i],
-                                                                                             desc.coltypes[i].cql_parameterized_type(),
-                                                                                             str(e)))
-            # Insert new object into tuple
-            tuple_set(res, i, val)
+                # Insert new object into tuple
+                tuple_set(res, i, val)
+        except Exception as e:
+            raise DriverException('Failed decoding result column "%s" of type %s: %s' % (desc.colnames[i],
+                                                                                         desc.coltypes[i].cql_parameterized_type(),
+                                                                                         str(e)))
+
+        return res
```
Confirmed and fixed, with one refinement over the literal suggestion.

Verified the claim is real: `ParseDesc.__init__` (`cassandra/parsing.pyx`) had no validation of `deserializers` vs `colnames` length, and `TupleRowParser.unpack_plain_row`/`unpack_col_encrypted_row` (`cassandra/obj_parser.pyx`) both index `desc.deserializers[i]` for `i in range(rowsize)` where `rowsize = desc.rowsize = len(colnames)`, under `@cython.boundscheck(False)`. So a mismatch would indeed turn into an out-of-bounds read rather than a safe `IndexError`.

I also checked whether this is reachable in practice: the only production construction site is `cassandra/row_parser.pyx`, where `colnames`, `coltypes`, and `deserializers` (via `make_deserializers(column_types)`) are all derived 1:1 from the same `column_metadata` list, so lengths always agree today. The only other construction site is the test helper in `tests/unit/test_protocol.py`. So this is a real gap in the safety net for an invariant that "shouldn't" break, but has no code-level guarantee — worth the cheap guard.

Fix (amended into the `Optimize column_encryption_policy checks in Cython parsers` commit that added the boundscheck decorators): added a length check in `ParseDesc.__init__` that raises `ValueError` on mismatch, rather than a plain `assert`. I used `raise ValueError` instead of `assert` deliberately: `assert` statements are stripped when Python runs with `-O`, which would silently remove exactly the safety net this guard exists to provide, given the hot loop's `boundscheck` is already disabled.

Added `test_parse_desc_rejects_mismatched_deserializers_length` (in `CythonParserTest`, `tests/unit/test_protocol.py`) which constructs a `ParseDesc` with a deliberately short `deserializers` list and asserts it raises `ValueError`, and confirms construction with matching lengths does not raise. Confirmed this test fails against the pre-fix code and passes post-fix. Full `tests/unit/` suite (779 passed, 38 skipped) still green with the Cython extensions rebuilt.

mykaul commented on tests/unit/test_protocol.py:
```diff
@@ -555,3 +559,315 @@ def test_frames_without_features(self):
+        self.assertEqual(msg.parsed_rows[0][1], 'hello')
+
+        # Verify contains_column was called for each value (but policy existence check happens once)
+        # Should be called 4 times (2 rows × 2 columns)
```
Confirmed and fixed. Searched the whole file for the `×` (U+00D7) character and found exactly the two occurrences flagged (lines 658 and 684, both in the "Should be called 4 times (2 rows × 2 columns)" comment) — no others. Replaced both with ASCII `x`. Amended into the `Add unit tests for column_encryption_policy optimization` commit that introduced these comments. Verified with `grep -c "×" tests/unit/test_protocol.py` returning 0 after the fix, and the full `tests/unit/` suite still passes (779 passed, 38 skipped).
